Thread-safe compilation API

arkpar opened the pull request:
This makes `Module` implement `Send` and `Sync`, allowing instances to be created and executed in different threads.

Closes #777

alexcrichton commented on crates/jit/src/compiler.rs:
```diff
@@ -51,13 +55,16 @@ pub enum CompilationStrategy {
     signatures: SignatureRegistry,
     strategy: CompilationStrategy,
     cache_config: CacheConfig,
+    inner: RwLock<CompilerInner>,
```
I personally think that this probably isn't the way we want to parallelize `Module` and `Store` and such. This means that compilation is still entirely serialized and you can't compile multiple modules in parallel, one of the main usages of enabling parallelism. I think we need to work to continue to refactor and reimplement the guts of `Compiler` to make them `Send` and `Sync` and use `&self`.

For example the `trampoline_park` field is going away in https://github.com/bytecodealliance/wasmtime/pull/957, and I don't think that the switch to `AtomicPtr<T>` here is correct in that it's not ever modified, it's just a pointer.

The main next change is to move `CodeMemory` to per-module instances, and then mutability shouldn't be needed on most of the fields here.

alexcrichton commented on crates/jit/src/instantiate.rs:
```diff
@@ -45,11 +44,19 @@ pub enum SetupError {
+
+// This is safe because the pointers are not modified and are only returned
+// through a reference bound to RawCompiledModule instance
+unsafe impl Send for FinishedFunctions {}
```
Instead of unsafely asserting send/sync for `FinishedFunctions`, could a newtype wrapper around `*const [VMFunctionBody]` be made to narrow-the scope of send/sync? That way we're not unsafely asserting thread safety about `DefinedFuncIndex` or `BoxedSlice`

alexcrichton commented on crates/runtime/src/jit_int.rs:
```diff
@@ -16,6 +20,10 @@ const JIT_NOACTION: u32 = 0;
 const JIT_REGISTER_FN: u32 = 1;
 const JIT_UNREGISTER_FN: u32 = 2;
 
+lazy_static::lazy_static! {
```
I've worked pretty hard to remove `lazy_static` from most crates because it often behaves in susprising ways. I don't understand much about the GDB integration here, but I think it would be better to have an instanced lock and/or atomics to manage this.

yurydelendik replied:
gdb will use `__jit_debug_descriptor` (see static below) on this module startup and during reactiving on `__jit_debug_register_code` breakpoint AFAIK. The `__jit_debug_descriptor` is a global symbol for entire module.

arkpar replied:
Is it safe to assume there won't be any concurrent access here?

arkpar commented:
This PR mostly addresses the use case of compiling once and then executing compiled code in multiple threads, which is arguably more important than parallel compilation. And it achieves this with relatively small changeset, which should not interfere with refactoring towards full-blown parallel API later. If this may still be accepted, @pepyakin Is going to take over this PR, addressing issues with `AtomicPtr`, safe wrapping `*const [VMFunctionBody]` and merging with #957

pepyakin commented:
I am going to close this for the time being.
